File: go/kadane/152_max_product.go
package kadane

// maxProduct 乘积最大子数组
func maxProduct(nums []int) int {

	// Kadane
	maxEndingHere, minEndingHere := nums[0], nums[0]
	maxSoFar := nums[0]

	for i := 1; i < len(nums); i++ {

		oldMax := maxEndingHere
		// 存在 nums[i] 是个负数，相乘后，由最大值变为最小值，最小值变为最大值
		maxEndingHere = max(maxEndingHere*nums[i], nums[i], minEndingHere*nums[i])
		minEndingHere = min(minEndingHere*nums[i], nums[i], oldMax*nums[i])

		maxSoFar = max(maxSoFar, maxEndingHere)
	}
	return maxSoFar
}

File: rust/src/kadane/_152_max_product.rs

/// 乘积最大子数组
pub fn max_product(nums: Vec<i32>) -> i32 {
    let (mut max_ending_here, mut min_ending_here) = (nums[0], nums[0]);
    let mut max_so_far = nums[0];

    for i in 1..nums.len() {
        let old_max = max_ending_here;

        max_ending_here = nums[i]
            .max(max_ending_here * nums[i])
            .max(min_ending_here * nums[i]);

        min_ending_here = nums[i]
            .min(min_ending_here * nums[i])
            .min(old_max * nums[i]);

        max_so_far = max_so_far.max(max_ending_here);
    }

    max_so_far
}
